### PR TITLE
capture(cdc): prevents process exits when Etcd session is done

### LINKED ISSUE #4752: [CC scenario]  PD leader unavailable temporally makes CDC node exited
### What did you do?

- As title
- CDC node exited with different ERROR:
   1. [2022/02/27 14:00:02.475 +00:00] [WARN] [check.go:81] ["check TiKV version failed"] [error="[CDC:ErrGetAllStoresFailed]get stores from pd failed: rpc error: code = DeadlineExceeded desc = context deadline exceeded"] [errorVerbose="[CDC:ErrGetAllStoresFailed]get stores from pd failed: rpc error: code = DeadlineExceeded desc = context deadline exceeded\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/normalize.go:164\ngithub.com/pingcap/tiflow/pkg/errors.WrapError\n\tgithub.com/pingcap/tiflow/pkg/errors/helper.go:30\ngithub.com/pingcap/tiflow/pkg/version.CheckStoreVersion\n\tgithub.com/pingcap/tiflow/pkg/version/check.go:165\ngithub.com/pingcap/tiflow/pkg/version.CheckClusterVersion\n\tgithub.com/pingcap/tiflow/pkg/version/check.go:76\ngithub.com/pingcap/tiflow/cdc.(*Server).Run\n\tgithub.com/pingcap/tiflow/cdc/server.go:180\ngithub.com/pingcap/tiflow/pkg/cmd/server.(*options).run\n\tgithub.com/pingcap/tiflow/pkg/cmd/server/server.go:129\ngithub.com/pingcap/tiflow/pkg/cmd/server.NewCmdServer.func1\n\tgithub.com/pingcap/tiflow/pkg/cmd/server/server.go:277\ngithub.com/spf13/cobra.(*Command).execute\n\tgithub.com/spf13/cobra@v1.2.1/command.go:856\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\tgithub.com/spf13/cobra@v1.2.1/command.go:974\ngithub.com/spf13/cobra.(*Command).Execute\n\tgithub.com/spf13/cobra@v1.2.1/command.go:902\ngithub.com/pingcap/tiflow/pkg/cmd.Run\n\tgithub.com/pingcap/tiflow/pkg/cmd/cmd.go:50\nmain.main\n\tcommand-line-arguments/main.go:22\nruntime.main\n\truntime/proc.go:225\nruntime.goexit\n\truntime/asm_amd64.s:1371"]
[2022/02/27 14:00:02.686 +00:00] [ERROR] [server.go:131] ["run server"] [error="[CDC:ErrCheckClusterVersionFromPD]failed to request PD%!(EXTRA string=response status: 503 Service Unavailable)\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/normalize.go:164\ngithub.com/pingcap/tiflow/pkg/version.CheckPDVersion\n\tgithub.com/pingcap/tiflow/pkg/version/check.go:120\ngithub.com/pingcap/tiflow/pkg/version.CheckClusterVersion\n\tgithub.com/pingcap/tiflow/pkg/version/check.go:85\ngithub.com/pingcap/tiflow/cdc.(*Server).Run\n\tgithub.com/pingcap/tiflow/cdc/server.go:180\ngithub.com/pingcap/tiflow/pkg/cmd/server.(*options).run\n\tgithub.com/pingcap/tiflow/pkg/cmd/server/server.go:129\ngithub.com/pingcap/tiflow/pkg/cmd/server.NewCmdServer.func1\n\tgithub.com/pingcap/tiflow/pkg/cmd/server/server.go:277\ngithub.com/spf13/cobra.(*Command).execute\n\tgithub.com/spf13/cobra@v1.2.1/command.go:856\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\tgithub.com/spf13/cobra@v1.2.1/command.go:974\ngithub.com/spf13/cobra.(*Command).Execute\n\tgithub.com/spf13/cobra@v1.2.1/command.go:902\ngithub.com/pingcap/tiflow/pkg/cmd.Run\n\tgithub.com/pingcap/tiflow/pkg/cmd/cmd.go:50\nmain.main\n\tcommand-line-arguments/main.go:22\nruntime.main\n\truntime/proc.go:225\nruntime.goexit\n\truntime/asm_amd64.s:1371"]
Error: run server: [CDC:ErrCheckClusterVersionFromPD]failed to request PD%!(EXTRA string=response status: 503 Service Unavailable)
2. [2022/02/27 14:01:39.420 +00:00] [INFO] [processor.go:294] ["processor exited"] [capture=db-cdc-ticdc-7.db-cdc-ticdc-peer.tidb1373933076657975328.svc:8301] [changefeed=362e4d46-8380-4f45-b103-5a6f040adfdb]
[2022/02/27 14:01:39.420 +00:00] [WARN] [etcd_worker.go:242] ["EtcdWorker reactor tick took too long"] [duration=1m13.411839507s]
[2022/02/27 14:01:39.420 +00:00] [WARN] [capture.go:459] ["session is disconnected"] [error="[CDC:ErrEtcdSessionDone]the etcd session is done"] [errorVerbose="[CDC:ErrEtcdSessionDone]the etcd session is done\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/normalize.go:164\ngithub.com/pingcap/tiflow/pkg/orchestrator.(*EtcdWorker).Run\n\tgithub.com/pingcap/tiflow/pkg/orchestrator/etcd_worker.go:160\ngithub.com/pingcap/tiflow/cdc/capture.(*Capture).runEtcdWorker\n\tgithub.com/pingcap/tiflow/cdc/capture/capture.go:450\ngithub.com/pingcap/tiflow/cdc/capture.(*Capture).run.func3\n\tgithub.com/pingcap/tiflow/cdc/capture/capture.go:316\nruntime.goexit\n\truntime/asm_amd64.s:1371"]
[2022/02/27 14:01:39.421 +00:00] [INFO] [capture.go:317] ["the processor routine has exited"] [error="[CDC:ErrCaptureSuicide]capture suicide"] [errorVerbose="[CDC:ErrCaptureSuicide]capture suicide\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/normalize.go:164\ngithub.com/pingcap/tiflow/cdc/capture.(*Capture).runEtcdWorker\n\tgithub.com/pingcap/tiflow/cdc/capture/capture.go:460\ngithub.com/pingcap/tiflow/cdc/capture.(*Capture).run.func3\n\tgithub.com/pingcap/tiflow/cdc/capture/capture.go:316\nruntime.goexit\n\truntime/asm_amd64.s:1371"]
[2022/02/27 14:01:39.510 +00:00] [INFO] [capture.go:254] ["the capture routine has exited"]
[2022/02/27 14:01:39.512 +00:00] [ERROR] [http_status.go:74] ["http server error"] [error="[CDC:ErrServeHTTP]serve http error: mux: server closed"] [errorVerbose="[CDC:ErrServeHTTP]serve http error: mux: server closed\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/normalize.go:164\ngithub.com/pingcap/tiflow/pkg/errors.WrapError\n\tgithub.com/pingcap/tiflow/pkg/errors/helper.go:30\ngithub.com/pingcap/tiflow/cdc.(*Server).startStatusHTTP.func1\n\tgithub.com/pingcap/tiflow/cdc/http_status.go:74\nruntime.goexit\n\truntime/asm_amd64.s:1371"]
[2022/02/27 14:01:39.518 +00:00] [ERROR] [client.go:752] ["[pd] fetch pending tso requests error"] [dc-location=global] [error="[PD:client:ErrClientGetTSO]context canceled: context canceled"]
[2022/02/27 14:01:39.520 +00:00] [INFO] [client.go:666] ["[pd] exit tso dispatcher"] [dc-location=global]
[2022/02/27 14:01:39.530 +00:00] [ERROR] [server.go:131] ["run server"] [error="[CDC:ErrNotOwner]this capture is not a owner\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/normalize.go:164\ngithub.com/pingcap/tiflow/pkg/etcd.CDCEtcdClient.GetOwnerRevision\n\tgithub.com/pingcap/tiflow/pkg/etcd/etcd.go:574\ngithub.com/pingcap/tiflow/cdc/capture.(*Capture).campaignOwner\n\tgithub.com/pingcap/tiflow/cdc/capture/capture.go:395\ngithub.com/pingcap/tiflow/cdc/capture.(*Capture).run.func2\n\tgithub.com/pingcap/tiflow/cdc/capture/capture.go:292\nruntime.goexit\n\truntime/asm_amd64.s:1371\nowner exited with error"]
[2022/02/27 14:01:39.532 +00:00] [ERROR] [client.go:752] ["[pd] fetch pending tso requests error"] [dc-location=global] [error="[PD:client:ErrClientGetTSO]context canceled: context canceled"]
[2022/02/27 14:01:39.532 +00:00] [INFO] [client.go:666] ["[pd] exit tso dispatcher"] [dc-location=global]
Error: run server: owner exited with error: [CDC:ErrNotOwner]this capture is not a owner
run server: owner exited with error: [CDC:ErrNotOwner]this capture is not a owner
3. [2022/03/02 06:27:14.699 +00:00] [INFO] [capture.go:254] ["the capture routine has exited"]
[2022/03/02 06:27:14.699 +00:00] [ERROR] [http_status.go:74] ["http server error"] [error="[CDC:ErrServeHTTP]serve http error: mux: server closed"] [errorVerbose="[CDC:ErrServeHTTP]serve http error: mux: server closed\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/normalize.go:164\ngithub.com/pingcap/tiflow/pkg/errors.WrapError\n\tgithub.com/pingcap/tiflow/pkg/errors/helper.go:30\ngithub.com/pingcap/tiflow/cdc.(*Server).startStatusHTTP.func1\n\tgithub.com/pingcap/tiflow/cdc/http_status.go:74\nruntime.goexit\n\truntime/asm_amd64.s:1371"]
[2022/03/02 06:27:14.699 +00:00] [ERROR] [client.go:752] ["[pd] fetch pending tso requests error"] [dc-location=global] [error="[PD:client:ErrClientGetTSO]context canceled: context canceled"]
[2022/03/02 06:27:14.699 +00:00] [INFO] [client.go:666] ["[pd] exit tso dispatcher"] [dc-location=global]
[2022/03/02 06:27:14.699 +00:00] [ERROR] [server.go:131] ["run server"] [error="rpc error: code = Unavailable desc = transport is closing\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/errors.go:174\ngithub.com/pingcap/errors.Trace\n\tgithub.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/juju_adaptor.go:15\ngithub.com/pingcap/tiflow/pkg/orchestrator.(*EtcdWorker).commitChangedState\n\tgithub.com/pingcap/tiflow/pkg/orchestrator/etcd_worker.go:409\ngithub.com/pingcap/tiflow/pkg/orchestrator.(*EtcdWorker).applyPatchGroups\n\tgithub.com/pingcap/tiflow/pkg/orchestrator/etcd_worker.go:337\ngithub.com/pingcap/tiflow/pkg/orchestrator.(*EtcdWorker).Run\n\tgithub.com/pingcap/tiflow/pkg/orchestrator/etcd_worker.go:207\ngithub.com/pingcap/tiflow/cdc/capture.(*Capture).runEtcdWorker\n\tgithub.com/pingcap/tiflow/cdc/capture/capture.go:450\ngithub.com/pingcap/tiflow/cdc/capture.(*Capture).campaignOwner\n\tgithub.com/pingcap/tiflow/cdc/capture/capture.go:428\ngithub.com/pingcap/tiflow/cdc/capture.(*Capture).run.func2\n\tgithub.com/pingcap/tiflow/cdc/capture/capture.go:292\nruntime.goexit\n\truntime/asm_amd64.s:1371\nowner exited with error"]
Error: run server: owner exited with error: rpc error: code = Unavailable desc = transport is closing
[2022/03/02 06:27:14.699 +00:00] [ERROR] [client.go:752] ["[pd] fetch pending tso requests error"] [dc-location=global] [error="[PD:client:ErrClientGetTSO]context canceled: context canceled"]

### What did you expect to see?

_No response_

### What did you see instead?

Sometimes one of cdc node exited;
Sometimes 7/8 cdc node exited.

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
5.4.0
```

Upstream TiKV version (execute `tikv-server --version`):

```console
5.4.0
```

TiCDC version (execute `cdc version`):

```console
5.4.0
```

### PR DESCRIPTION
**Pull request by liuzix**
### What problem does this PR solve?

Issue Number: close #4752

### What is changed and how it works?
- Do not trigger a process exit if when the error has happened, the Etcd session is done. This is a temporary fix to prevent the process from exiting when the error is caused by a PD(Etcd) availability issue. This is not a fix-all for all related cases.

### Check List 

Tests 

 - Integration test
 - Manual test

Passed manual tests with injecting network partition between TiCDC nodes and PD one by one.

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

### Release note 

```release-note
Mitigate risk of process exiting when PD is not available.
```

### CODE REVIEW

File: cdc/capture/capture.go [hunk 1]
```diff
@@ -260,6 +260,17 @@ func (c *Capture) Run(ctx context.Context) error {
 			log.Info("capture recovered", zap.String("captureID", c.info.ID))
 			continue
 		}
+		select {
+		case <-c.session.Done():
+			log.Warn("capture encountered error while session is gone. Retrying later.",
```
Comment (CharlesCheung96) [inline cdc/capture/capture.go:265]: 1. Should it be written as "capture encountered error while session is **done**"?
2. Can we throw the session done [here](https://github.com/pingcap/tiflow/blob/d141ee67f015377cfc288e5ea68c101614eaed16/cdc/capture/capture.go#L471) and handle it uniformly.

Comment (liuzix) [reply cdc/capture/capture.go:265]: 1. **gone** feels equally okay to me.
2. This error might not be the immediate cause of the problem, so we need to consider all types of errors that can be caused by the PD cluster being unavailable.

**Pull request closed without merging**